### PR TITLE
added token utility

### PR DESCRIPTION
**Pull request by ThomasLiARDJAVA**
# Administrative Info

## Changes
What changes did you make?
- Added token utility functions, `generateToken`, and `validateToken`.
- Started `POST /login` route
- Added comments for parts assigned to Aditya
## Dependencies/Library installed: 
- Json-web-token

## Checklist

Remove any items which are not applicable.

- [x] I have performed a self-review of my own code
- [x] I have added tested my program thoroughly
- [x] Did you add/sub fields in the environment file? if so please update .env_sample

### CODE REVIEW

File: app/auth/token.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+const jwt = require('jsonwebtoken');
+
+/**
+ * validates the token, and return the user json.
+ * @param token
+ * @returns {Promise<user json>}
+ */
+async function validateToken(token){
+    return jwt.verify(token, process.env.SECRETKEY, async function(err, authData) {
+        if(err) {
```
Comment (rcherukuru2000) [inline app/auth/token.js:10]: Codacy has a fix for the issue: [Expected space(s) after "if".](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
        if (err) {
```

File: app/auth/token.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            return null;
+        } else {
+            return authData;
```
Comment (rcherukuru2000) [inline app/auth/token.js:13]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
      return authData;
```

File: app/auth/token.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        }
+    });
+}
+
+/**
+ * generate a token with the given user.
+ * @param userDataPacket
+ * @returns {Promise<token|*>}
```
Comment (rcherukuru2000) [inline app/auth/token.js:21]: Codacy has a fix for the issue: [Use @return instead.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
 * @return {Promise<token|*>}
```

File: app/auth/token.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * @returns {Promise<user json>}
+ */
+async function validateToken(token){
+    return jwt.verify(token, process.env.SECRETKEY, async function(err, authData) {
+        if(err) {
+            return null;
+        } else {
+            return authData;
+        }
+    });
```
Comment (rcherukuru2000) [inline app/auth/token.js:15]: Codacy has a fix for the issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
  });
```

File: app/auth/token.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ * generate a token with the given user.
+ * @param userDataPacket
+ * @returns {Promise<token|*>}
+ */
+async function generateToken(user){
+    return jwt.sign(user, process.env.SECRETKEY);
+}
+
+module.exports = {
+    validateToken,
```
Comment (rcherukuru2000) [inline app/auth/token.js:28]: Codacy has a fix for the issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
  validateToken,
```

File: app/auth/token.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * validates the token, and return the user json.
+ * @param token
+ * @returns {Promise<user json>}
+ */
+async function validateToken(token){
+    return jwt.verify(token, process.env.SECRETKEY, async function(err, authData) {
+        if(err) {
+            return null;
```
Comment (rcherukuru2000) [inline app/auth/token.js:11]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
      return null;
```

File: app/auth/token.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        } else {
```
Comment (rcherukuru2000) [inline app/auth/token.js:12]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
    } else {
```

File: app/auth/token.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+const jwt = require('jsonwebtoken');
+
+/**
+ * validates the token, and return the user json.
+ * @param token
+ * @returns {Promise<user json>}
+ */
+async function validateToken(token){
+    return jwt.verify(token, process.env.SECRETKEY, async function(err, authData) {
```
Comment (rcherukuru2000) [inline app/auth/token.js:9]: Codacy has a fix for the issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
  return jwt.verify(token, process.env.SECRETKEY, async function(err, authData) {
```

File: app/auth/token.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        if(err) {
+            return null;
+        } else {
+            return authData;
+        }
```
Comment (rcherukuru2000) [inline app/auth/token.js:14]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
    }
```

File: app/auth/token.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    });
+}
+
+/**
+ * generate a token with the given user.
+ * @param userDataPacket
+ * @returns {Promise<token|*>}
+ */
+async function generateToken(user){
+    return jwt.sign(user, process.env.SECRETKEY);
```
Comment (rcherukuru2000) [inline app/auth/token.js:24]: Codacy has a fix for the issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
  return jwt.sign(user, process.env.SECRETKEY);
```

File: app/auth/token.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+module.exports = {
+    validateToken,
+    generateToken
```
Comment (rcherukuru2000) [inline app/auth/token.js:29]: Codacy has a fix for the issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
  generateToken
```

File: app/routes/index.js [hunk 12]
```diff
@@ -14,7 +14,29 @@ router.get('/login', async function(req, res, next){
 });
 
 router.post('/login', async function(req, res, next){
-  // TODO: fill in login logic
+  try {
+
+    const user = req.body;
+    const user_in_db = await usersModel.getByEmail(user.email);
+
+    if(user_in_db.length === 0){
```
Comment (rcherukuru2000) [inline app/routes/index.js:22]: Codacy has a fix for the issue: [Expected space(s) after "if".](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
    if (user_in_db.length === 0){
```

File: app/auth/token.js (hunk 1 above)
Comment (rcherukuru2000) [inline app/auth/token.js:10]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
    if(err) {
```

File: app/auth/token.js [hunk 13]
```diff
@@ -0,0 +1,30 @@
+const jwt = require('jsonwebtoken');
+
+/**
+ * validates the token, and return the user json.
+ * @param token
+ * @returns {Promise<user json>}
+ */
+async function validateToken(token){
+  return jwt.verify(token, process.env.SECRETKEY, async function(err, authData) {
+        if (err) {
```
Comment (rcherukuru2000) [inline app/auth/token.js:10]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
    if(err) {
```

File: app/auth/token.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
+const jwt = require('jsonwebtoken');
+
+/**
+ * validates the token, and return the user json.
+ * @param token
+ * @returns {Promise<user json>}
+ */
+async function validateToken(token){
+  return jwt.verify(token, process.env.SECRETKEY, async function(err, authData) {
+    if(err) {
```
Comment (rcherukuru2000) [inline app/auth/token.js:10]: Codacy has a fix for the issue: [Expected space(s) after "if".](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
        if (err) {
```

File: app/auth/token.js (hunk 13 above)
Comment (rcherukuru2000) [inline app/auth/token.js:10]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
    if(err) {
```

File: app/middlewares/verifyCookieToken.js [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+ * @param res
+ * @param next
+ * @returns {Promise<void>}
+ */
+async function verifyCookieToken(req, res, next) {
+    try {
+        const token = req.cookies.token;
+        const payload = token ? await tokenUtil.validateToken(token) : null;
+        if (payload) {
+            req.user = payload.user;
```
Comment (rcherukuru2000) [inline app/middlewares/verifyCookieToken.js:15]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
      req.user = payload.user;
```

File: app/routes/index.js [hunk 16]
```diff
@@ -14,7 +16,36 @@ router.get('/login', async function(req, res, next){
...
+    const user_in_db = await usersModel.getByEmail(user.email);
+
+    if (user_in_db.length === 0){
+      // eventually need to make a better page and redirect them to signup.
+      return res.status(401)
+          .json({status: 'failed', message: 'Email not found, user does not exist in database'});
+    }
+
+    // add password comparison bcrypt.compareSync(user.password, user_in_db[0].password) returns boolean
+    if(!bcrypt.compareSync(user.password, user_in_db[0].password)){
```
Comment (rcherukuru2000) [inline app/routes/index.js:31]: Codacy has a fix for the issue: [Expected space(s) after "if".](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
    if (!bcrypt.compareSync(user.password, user_in_db[0].password)){
```

File: app/middlewares/verifyCookieToken.js [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * verifies cookie token for each parking spot, and returns info back from cookie to req.spotInfo.
+ * @param req
+ * @param res
+ * @param next
+ * @returns {Promise<void>}
+ */
+async function verifyCookieToken(req, res, next) {
+    try {
+        const token = req.cookies.token;
```
Comment (rcherukuru2000) [inline app/middlewares/verifyCookieToken.js:12]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
    const token = req.cookies.token;
```

File: app/middlewares/verifyCookieToken.js [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+        const payload = token ? await tokenUtil.validateToken(token) : null;
+        if (payload) {
+            req.user = payload.user;
+            next();
+        } else {
+            res.status(401)
+                .json({status: 'failed', data: 'Session over'});
+        }
+    } catch (err){
+        res.status(502)
```
Comment (rcherukuru2000) [inline app/middlewares/verifyCookieToken.js:22]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
    res.status(502)
```

File: app/middlewares/verifyCookieToken.js [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+ * verifies cookie token for each parking spot, and returns info back from cookie to req.spotInfo.
+ * @param req
+ * @param res
+ * @param next
+ * @returns {Promise<void>}
+ */
+async function verifyCookieToken(req, res, next) {
+    try {
+        const token = req.cookies.token;
+        const payload = token ? await tokenUtil.validateToken(token) : null;
```
Comment (rcherukuru2000) [inline app/middlewares/verifyCookieToken.js:13]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
    const payload = token ? await tokenUtil.validateToken(token) : null;
```

File: app/middlewares/verifyCookieToken.js [hunk 20]
```diff
@@ -0,0 +1,27 @@
+const tokenUtil = require('../auth/token');
+
+/**
+ * verifies cookie token for each parking spot, and returns info back from cookie to req.spotInfo.
+ * @param req
+ * @param res
+ * @param next
+ * @returns {Promise<void>}
```
Comment (rcherukuru2000) [inline app/middlewares/verifyCookieToken.js:8]: Codacy has a fix for the issue: [Use @return instead.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
 * @return {Promise<void>}
```

File: app/middlewares/verifyCookieToken.js [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+ */
+async function verifyCookieToken(req, res, next) {
+    try {
+        const token = req.cookies.token;
+        const payload = token ? await tokenUtil.validateToken(token) : null;
+        if (payload) {
+            req.user = payload.user;
+            next();
+        } else {
+            res.status(401)
```
Comment (rcherukuru2000) [inline app/middlewares/verifyCookieToken.js:18]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
      res.status(401)
```

File: app/middlewares/verifyCookieToken.js [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+                .json({status: 'failed', data: 'Session over'});
+        }
+    } catch (err){
```
Comment (rcherukuru2000) [inline app/middlewares/verifyCookieToken.js:21]: Codacy has a fix for the issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
  } catch (err){
```

File: app/middlewares/verifyCookieToken.js [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+        res.status(502)
+            .json({status: 'failed', data: 'malfunction'});
```
Comment (rcherukuru2000) [inline app/middlewares/verifyCookieToken.js:23]: Codacy has a fix for the issue: [Expected indentation of 8 spaces but found 12.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
        .json({status: 'failed', data: 'malfunction'});
```

File: app/middlewares/verifyCookieToken.js [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+async function verifyCookieToken(req, res, next) {
+    try {
+        const token = req.cookies.token;
+        const payload = token ? await tokenUtil.validateToken(token) : null;
+        if (payload) {
+            req.user = payload.user;
+            next();
+        } else {
+            res.status(401)
+                .json({status: 'failed', data: 'Session over'});
```
Comment (rcherukuru2000) [inline app/middlewares/verifyCookieToken.js:19]: Codacy has a fix for the issue: [Expected indentation of 10 spaces but found 16.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
          .json({status: 'failed', data: 'Session over'});
```

File: app/middlewares/verifyCookieToken.js [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+ * @param req
+ * @param res
+ * @param next
+ * @returns {Promise<void>}
+ */
+async function verifyCookieToken(req, res, next) {
+    try {
+        const token = req.cookies.token;
+        const payload = token ? await tokenUtil.validateToken(token) : null;
+        if (payload) {
```
Comment (rcherukuru2000) [inline app/middlewares/verifyCookieToken.js:14]: Codacy has a fix for the issue: [Expected indentation of 4 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group17/cse110-fa21-group17/pullRequest?prid=8443778)
```suggestion
    if (payload) {
```

**Review by adityaamoghe — COMMENTED** [review]
Looks good!

**Review by s8sun — APPROVED** [review]

**Pull request merged**